en/lessons/basics/modules.md

```diff
@@ -1,316 +1,195 @@
-  defstruct name: "Sean", roles: []
-end
-```
-
-Let's create some structs:
-
-```elixir
-iex> %Example.User{}
-%Example.User<name: "Sean", roles: [], ...>
-
-iex> %Example.User{name: "Steve"}
-%Example.User<name: "Steve", roles: [], ...>
-
-iex> %Example.User{name: "Steve", roles: [:manager]}
-%Example.User<name: "Steve", roles: [:manager]>
-```
+id :: a -> a
+id a = a
 
-We can update our struct just like we would a map:
-
-```elixir
-iex> steve = %Example.User{name: "Steve"}
-%Example.User<name: "Steve", roles: [...], ...>
-iex> sean = %{steve | name: "Sean"}
-%Example.User<name: "Sean", roles: [...], ...>
+compose :: (b -> c) -> (a -> b) -> (a -> c)
+compose f g = \a -> f (g a)
 ```
 
-Most importantly, you can match structs against maps:
+Here is the same module, but residing at a different filepath.
 
-```elixir
-iex> %{name: "Sean"} = sean
-%Example.User<name: "Sean", roles: [...], ...>
-```
+```haskell
+-- src/Data/Examples/Example.hs
+module Data.Examples.Example where
 
-As of Elixir 1.8 structs include custom introspection.
-To understand what this means and how we are to use it let us inspect our `sean` capture:
+id :: a -> a
+id a = a
 
-```elixir
-iex> inspect(sean)
-"%Example.User<name: \"Sean\", roles: [...], ...>"
+compose :: (b -> c) -> (a -> b) -> (a -> c)
+compose f g = \a -> f (g a)
 ```
 
-All of our fields are present which is okay for this example but what if we had a protected field we didn't want to include?
-The new `@derive` feature let's us accomplish just this!
-Let's update our example so `roles` are no longer included in our output:
+Modules can export a subset of the entities they contain. In the example below
+only the type `Identity` (and its constructor) and `id` are exported, `compose`
+is completely isolated to the local module scope.
 
-```elixir
-defmodule Example.User do
-  @derive {Inspect, only: [:name]}
-  defstruct name: nil, roles: []
-end
-```
+```haskell
+-- src/Identity.hs
+module Identity (
+  Identity(..),
+  id
+) where
 
-_Note_: we could also use `@derive {Inspect, except: [:roles]}`, they are equivalent.
+data Identity a = Identity a
 
-With our updated module in place let's take a look at what happens in `iex`:
+id :: a -> a
+id a = a
 
-```elixir
-iex> sean = %Example.User{name: "Sean"}
-%Example.User<name: "Sean", ...>
-iex> inspect(sean)
-"%Example.User<name: \"Sean\", ...>"
+compose :: (b -> c) -> (a -> b) -> (a -> c)
+compose f g = \a -> f (g a)
 ```
 
-The `roles` are excluded from output!
-
-## Composition
-
-Now that we know how to create modules and structs let's learn how to add existing functionality to them via composition.
-Elixir provides us with a variety of different ways to interact with other modules.
-
-### `alias`
+Modules can also export types, but hide constructors. This means (with respect
+to the example below) that you can speak about things of type `Identity` but you
+cannot construct them yourself. These are sometimes referred to as opaque types.
```

Comment:
Similarly, it's difficult to decide whether this should go in this lesson. It would definitely be helpful to mention opaque types at some point in Haskell School though.